### DISCUSSION TITLE
additionalProperty or featureList, what to do?

### DISCUSSION THREAD

**Discussion by jvandriel** [schemaorg/schemaorg#3328] [Q&A]
While working on some code examples for somebody I ran into a bit of a conundrum regarding `SoftwareApplication`.

This type has a `featureList` property yet its expected value is either `Text` or `Url`, meaning the expected value doesn't allow for a type like `DefinedTerm` to be used so that one can express (for example) the `name` and `description` of a feature.

The alternative would be to specify a multi-typed entity like `["Product","SoftwareApplication"]` and use `additionalProperty > PropertyValue` instead. Yet this makes me wonder what the point of having `featureList` is, do we consider feature = property?

So what do you think would be the best solution here?

1. Expand the range of `featureList` to include `DefinedTerm` 
2. Create an MTE and use `additionalProperty > PropertyValue` to describe the individual features of a `SoftwareApplication`

**Comment by jvandriel** [comment]
Just noticed this question seems to overlap with issue #3250.

**Comment by jvandriel** [comment]
Another thing I'm wondering is how does `featureList` fit into describing the differences between variants of a `SoftwareApplication`? 

Many sites offer different feature bundles depending on the type of subscription you choose (and they often actually use the term _features_ for this). Though in my perception `ProductGroup` or `ProductModel` should be used here to describe the properties/features each variant has whereas each individual variant (`Product`) describes the properties/features supplemental to that.

Something we can't do when just using `SoftwareApplication` (or any of its sub types because it's not a subtype of `Product`) as `featureList` doesn't fit into this construct/model/whatever-its-called and possibly even conflicts when using an MTE like `["ProductModel","SoftwareApplication"]`. After all, `ProductGroup`/`ProductModel` + `additionalProperty` lets you specify differences between variants per property specifically  yet `featureList` currently, at best, provides a list of strings or urls representing a group of features/properties.

**Comment by danbri** [comment]
When you see “url or text”, that is a kind of schema.org way of saying “edge of the known universe”, ie. that there be more territory to map here, but in the meantime, don’t be surprised by urls. This effectively does something similar to DefinedTerm…

**Comment by jvandriel** [comment]
I get the point of _“edge of the known universe”_  but for the example I was working on there aren't any urls per feature I can refer to. As on many sites offering software subscriptions, this example has 1 page (eg. `ProductModel`) which mentions all their features (`PropertyValue.name`) and what they do (`PropertyValue.description`).

Now if I were to use `featureList` I would end up providing a list of comma separated string values (=`PropertyValue.name`) yet if I were to use `additionalProperty > PropertyValue` I can be much more expressive. Yet then the question becomes, can I semantically consider a **feature = property**?

And if that's considered OK, wouldn't that mute any argument for adding `DefinedTerm` to the range of `featureList`?

**Comment by RichardWallis** [comment]
The purpose of the `featureList` property seems be limited to a textual list directly entered or in a linked document.  What you seem to be asking for is more like a `feature` property which would more appropriately have `DefinedTerm` in its range.

However, I think you are just touching the surface of things you may want to describe about a `SoftwareApplication` that you currently can not: what audience is it targeted at, what is the brand, is it family friendly, etc., etc. 

Many of thes questions and associated discussions have lead to the current shape of the `Product` type.  So instead of reinventing the wheel, would it not be simpler to agree that a `SoftwareApplication` could also be a `Product` and take up the multi-typing suggestion made earlier.

**Comment by jvandriel** [comment]
I'd be all up for resolving this via multi-typing I'm just asking these questions here (and not as an issue) because I'm suffering from a brain-freeze caused by questions like:

1. Can I semantically consider a feature = a property? (I personally think the difference in wording is more a marketing thing than a semantical one)
2. Does the use of `featureList` become superfluous when using an MTE (while using `additionalProperty`), or is it still nice to have both methods? (`featureList` does offer a nice and simple method to express/extract a compacted version of that same info).

And just to be clear, I've been taking _“edge of the known universe”_ well into account and am not contemplating this for any specific so-called SEO purposes (although I might actually use it in real life). I was just wondering how I could express this within the boundaries of current schema.org vocab constraints without stretching things so far that I start introducing ambiguity as opposed to any additional meaning I'm hoping to express.

**Comment by ljgarcia** [comment]
@jvandriel what should be added to the range DefinedTerm and/or PropertyValue?

**Comment by jvandriel** [comment]
Nothing really @ljgarcia. I just noticed your discussion and was wondering (since it was mentioned in the discussion) what you wanted to do with `featureList` given the discussion was going about adding `DefinedTerm` to the range of some types.

And since I was struggling whether to use `featureList` or multi-typing the entity and use `additionalProperty` + `PropertyValue` instead I thought this discussion might be of influence to what you're discussing in #3250.

**Comment by mfhepp** [comment]
I am not sure I understand your application scenario completely, but as for product / services features, we basically have two options:

1. standardize them within schema.org; this is feasible only for characteristics that make sense for a large share of all products (like color, dimensions, manufacturer).
2. standardize them outside of schema.org, either as a Web vocabulary / ontology, or, simpler, by defining reliable, unique identifiers (e.g. "P1234" for "warp_speed") and then using the [schema:additionalProperty](https://schema.org/additionalProperty) mechanism **with these identifiers and a suitable prefix**.

This approach has been described in detail in [this paper](https://www.springerprofessional.de/integrating-product-classification-standards-into-schema-org-ecl/15418776), the [principle is explained in here](https://www.stalsoft.com/publication/odbase-meta4es-2017/).

The advantage of this approach is that (1) you can decouple the evolution of the standardization of product properties from the evolution of schema.org, (2) you avoid a lot of legal trouble in comparison to deriving Web vocabularies from other standards like eClass, UNSPSC, ETIM, etc. (I tried for years...)., and (3) you can always use the latest version of the referenced standard without waiting for schema.org updates.

/CC @danbri

**Comment by jvandriel** [comment]
I'm totally fine with option [2] @mfhepp. There was just some confusement on my end as to how to express the properties (a.k.a. features) of a `SoftwareApplication` because of the fact that Class has a `featureList` property.

Though I've come to the conclusion it is quite OK to be able to express features as a list of strings as sort of a simple summary. Whereas `additionalProperty` can be used to describe each property in depth (as you described).

**Comment by mfhepp** [comment]
For such „boolean“ product characteristics, you can also use the additionalProperty pattern with a value of schema:true.Am 01.08.2023 um 14:15 schrieb Jarno van Driel ***@***.***>:﻿
I'm totally fine with option [2] @mfhepp. There was just some confusement on my end as to how express the properties (a.k.a. features) of a SoftwareApplication because of the fact that Class has a featureList property.
Though I've come to the conclusion it is quite OK to be able to express features as a list of strings as sort of a simple summary. Whereas additionalProperty can be used to describe each property in depth (as you described).
